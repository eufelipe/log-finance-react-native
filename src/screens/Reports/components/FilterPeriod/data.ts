import {addDays, startOfMonth, startOfWeek} from 'date-fns';

export interface FilterPeriodItem {
  id: string;
  active?: boolean;
  interval: {
    start: Date;
    end: Date;
  };
}

const today = new Date();
const last15Days = addDays(today, -15);
const month = startOfMonth(today);
const week = startOfWeek(today);

const DATA: FilterPeriodItem[] = [
  {
    id: 'today',
    interval: {
      start: today,
      end: today,
    },
    active: true,
  },
  {
    id: 'week',
    interval: {
      start: week,
      end: today,
    },
  },
  {
    id: '15-days',
    interval: {
      start: last15Days,
      end: today,
    },
  },
  {
    id: 'month',
    interval: {
      start: month,
      end: today,
    },
  },
];

export default DATA;
