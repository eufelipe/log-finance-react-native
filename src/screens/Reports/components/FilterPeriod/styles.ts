import styled from 'styled-components/native';
import {Text, TouchableOpacity} from 'styles/layout';

export const Container = styled.View`
  flex-direction: row;
`;

export const FlatList = styled.FlatList`
  padding-top: ${({theme}) => theme.spacing.great}px;
`;

export const Touchable = styled(TouchableOpacity)`
  margin: 0px ${({theme}) => theme.spacing.great}px
    ${({theme}) => theme.spacing.tall}px;
`;

export const Label = styled(Text)<{active: boolean}>`
  font-family: ${({theme}) => theme.fontFamily.regular};
  font-size: ${({theme}) => theme.fontSizes.small};
  font-weight: ${({theme}) => theme.fontWeight.bold};
  color: ${({theme, active = false}) =>
    active ? theme.colors.white : theme.colors.gray};
`;
