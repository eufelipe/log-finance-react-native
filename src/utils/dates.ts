import {
  format as formatPattern,
  isValid,
  parseISO as parseISOFNS,
} from 'date-fns';

import locale from 'date-fns/locale/pt-BR';

export const DATE_FORMAT_ISO = 'yyyy-MM-dd';
export const DATE_FORMAT_BR = 'dd/MM/yyyy';
export const DATE_FORMAT_FRIENDLY = " dd/MM 'às' HH:mm";
export const DATE_FORMAT_FRIENDLY_TIME = "'às' HH:mm";

export function parseStringToDate(value: string): Date {
  const parsedDate = parseISOFNS(value);
  if (!isValid(parsedDate)) {
    throw Error('canot parse date');
  }
  return parsedDate;
}
export function parseDateToString({
  value,
  format = DATE_FORMAT_BR,
}: {
  value: Date;
  format?: string;
}): string {
  return formatPattern(value, format, {locale});
}

export const getDate = (date: Date): string => {
  return formatPattern(date, DATE_FORMAT_ISO, {locale});
};

export const getDateToday = (): string => {
  const today = new Date();
  return formatPattern(today, DATE_FORMAT_ISO, {locale});
};

export const getDateAndHourFriendly = (value: Date): string => {
  return formatPattern(value, DATE_FORMAT_FRIENDLY, {locale});
};

export const getTimeFriendly = (value: Date): string => {
  return formatPattern(value, DATE_FORMAT_FRIENDLY_TIME, {locale});
};
