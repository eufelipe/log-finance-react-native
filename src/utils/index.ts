export {default as number} from './number';
export {default as strings} from './strings';
