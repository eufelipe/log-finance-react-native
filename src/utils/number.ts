export const sanitizeToFloat = (value: string): number => {
  value = value.replace(/[^0-9.]+/, '');
  if (value.includes('.') && value.includes(',')) {
    value = value.replace(/\./g, '').replace(',', '.');
  } else if (value.includes(',')) {
    value = value.replace(',', '.');
  }
  return parseFloat(value) || 0;
};
