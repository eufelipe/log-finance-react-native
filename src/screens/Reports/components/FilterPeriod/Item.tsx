import React from 'react';
import {useTranslation} from 'react-i18next';
import {FilterPeriodItem} from './data';

import {Touchable, Label} from './styles';

interface ItemProps {
  item: FilterPeriodItem;
  active?: boolean;
  onPress: (item: FilterPeriodItem) => void;
}

const Item = ({item, onPress}: ItemProps): JSX.Element => {
  const {t} = useTranslation('reports');

  return (
    <Touchable onPress={() => onPress(item)}>
      <Label active={!!item.active}>{t(item.id)}</Label>
    </Touchable>
  );
};

export default Item;
