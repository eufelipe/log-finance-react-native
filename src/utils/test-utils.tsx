/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react';
import {render, RenderAPI, RenderOptions} from '@testing-library/react-native';

import {theme} from 'styles/theme';
import {ThemeProvider} from 'styled-components/native';

import {NavigationContainer} from '@react-navigation/native';

interface Props {
  children: JSX.Element;
}

const AllTheProviders = ({children}: Props) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any>,
  options?: RenderOptions,
): RenderAPI => render(ui, {wrapper: AllTheProviders, ...options});

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
