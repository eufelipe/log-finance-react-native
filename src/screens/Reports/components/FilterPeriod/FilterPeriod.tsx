import React, {useCallback, useState} from 'react';

import Item from './Item';

import {Container, FlatList} from './styles';

import DATA, {FilterPeriodItem} from './data';

interface Props {
  onChangeFilter: (filterPeriod: FilterPeriodItem) => void;
}

const FilterPeriod = ({onChangeFilter}: Props): JSX.Element => {
  const [first] = DATA;

  const [data, setData] = useState(DATA ?? []);

  const [, setFilterPeriodSelected] = useState<FilterPeriodItem>(first);

  const listKeyExtractor = useCallback(item => item.id, []);

  const onPress = useCallback(
    (filterPeriod: FilterPeriodItem) => {
      setFilterPeriodSelected(filterPeriod);

      const merge = data.map(item =>
        item.id === filterPeriod.id
          ? {...filterPeriod, active: true}
          : {...item, active: false},
      );

      setData(merge);
      onChangeFilter(filterPeriod);
    },
    [data, onChangeFilter],
  );

  const renderItem = useCallback(
    ({item}) => <Item item={item} onPress={onPress} />,
    [onPress],
  );

  return (
    <Container>
      <FlatList
        data={data}
        horizontal
        renderItem={renderItem}
        keyExtractor={listKeyExtractor}
      />
    </Container>
  );
};

export default FilterPeriod;
