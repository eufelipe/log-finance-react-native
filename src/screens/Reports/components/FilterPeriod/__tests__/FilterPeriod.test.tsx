/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';

import {render} from 'utils/test-utils';
import FilterPeriod from '../FilterPeriod';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({t: (key: any) => key}),
}));

const onChangeFilter = jest.fn();

describe('FilterPeriod', () => {
  it('Test match snapshot FilterPeriod', () => {
    // given
    const props = {
      onChangeFilter,
    };

    // when
    const {toJSON} = render(<FilterPeriod {...props} />);

    // then
    expect(toJSON()).toMatchSnapshot();
  });
});
