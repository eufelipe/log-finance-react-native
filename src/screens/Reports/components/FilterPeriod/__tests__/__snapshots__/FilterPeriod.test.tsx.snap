// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPeriod Test match snapshot FilterPeriod 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
    ]
  }
>
  <RCTScrollView
    data={
      Array [
        Object {
          "active": true,
          "id": "today",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-11-26T02:59:59.999Z,
          },
        },
        Object {
          "id": "week",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-11-21T03:00:00.000Z,
          },
        },
        Object {
          "id": "15-days",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-11-12T02:59:59.999Z,
          },
        },
        Object {
          "id": "month",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-11-01T03:00:00.000Z,
          },
        },
        Object {
          "id": "quarter",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-10-01T03:00:00.000Z,
          },
        },
        Object {
          "id": "semester",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-05-31T02:59:59.999Z,
          },
        },
        Object {
          "id": "year",
          "interval": Object {
            "end": 2021-11-27T02:59:59.999Z,
            "start": 2021-01-01T03:00:00.000Z,
          },
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "paddingTop": 21.428571428571427,
        },
      ]
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={true}
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            today
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            week
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            15-days
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            month
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            quarter
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            semester
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 42.857142857142854,
              "marginLeft": 21.428571428571427,
              "marginRight": 21.428571428571427,
              "marginTop": 0,
              "opacity": 1,
            }
          }
        >
          <Text
            active={false}
            style={
              Array [
                Object {
                  "color": "#464E5F",
                  "fontFamily": "Poppins-Regular",
                  "fontSize": 28,
                  "fontWeight": "700",
                },
              ]
            }
          >
            year
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
